src/Orleans.Runtime/Placement/ResourceOptimizedPlacementDirector.cs

```diff
@@ -0,0 +1,327 @@
+            {
+                relevantSilos.Add(new(silo, stats));
+            }
+        }
+
+        int chooseFrom = (int)Math.Ceiling(Math.Sqrt(relevantSilos.Count));
+        Dictionary<SiloAddress, float> chooseFromSilos = [];
+
+        while (chooseFromSilos.Count < chooseFrom)
+        {
+            int index = Random.Shared.Next(relevantSilos.Count);
+            var pickedSilo = relevantSilos[index];
+
+            relevantSilos.RemoveAt(index);
+
+            float score = CalculateScore(pickedSilo.Value);
+            chooseFromSilos.Add(pickedSilo.Key, score);
+        }
+
+        var orderedByLowestScore = chooseFromSilos.OrderBy(kv => kv.Value);
+
+        // there could be more than 1 silo that has the same score, we pick 1 of them randomly so that we dont continuously pick the first one.
+        var lowestScore = orderedByLowestScore.First().Value;
+        var shortListedSilos = orderedByLowestScore.TakeWhile(p => p.Value == lowestScore).ToList();
+        var winningSilo = shortListedSilos[Random.Shared.Next(shortListedSilos.Count)];
+
+        return winningSilo;
+    }
+
+    private KeyValuePair<SiloAddress, float> GetBestSiloCandidate_V2(SiloAddress[] compatibleSilos)
+    {
+        KeyValuePair<int, float> pick;
+
+        int compatibleSilosCount = compatibleSilos.Length;
+        if (compatibleSilosCount * Unsafe.SizeOf<KeyValuePair<int, ResourceStatistics>>() <= OneKiloByte)  // it is good practice not to allocate more than 1 kilobyte of memory on the stack
+        {
+            pick = MakePick(stackalloc KeyValuePair<int, ResourceStatistics>[compatibleSilosCount]);
+        }
+        else
+        {
+            var relevantSilos = ArrayPool<KeyValuePair<int, ResourceStatistics>>.Shared.Rent(compatibleSilosCount);
+            pick = MakePick(relevantSilos.AsSpan());
+            ArrayPool<KeyValuePair<int, ResourceStatistics>>.Shared.Return(relevantSilos);
+        }
+
+        foreach (var silo in compatibleSilos)
+        {
+            if (silo.GetConsistentHashCode() == pick.Key)
+            {
+                return new KeyValuePair<SiloAddress, float>(silo, pick.Value);
+            }
+        }
+
+        // It should never come to this point, unless 'GetConsistentHashCode' isnt consistent, which if its the case,
+        // this code can act as a 'tester' for that. This would be exceptional, so its better to stop the program.
+        throw new InvalidOperationException("No hash code from the list of compatible silos matched the picked silo's hash code.");
+
+        KeyValuePair<int, float> MakePick(Span<KeyValuePair<int, ResourceStatistics>> relevantSilos)
+        {
+            int relevantSilosCount = 0;
+            foreach (var silo in compatibleSilos)
+            {
+                if (_siloStatistics.TryGetValue(silo, out var stats) && !stats.IsOverloaded)
+                {
+                    relevantSilos[relevantSilosCount++] = new(silo.GetConsistentHashCode(), stats);
```

Comment:
Instead of using the hash code of the silo, we can use the index into the `compatibleSilos` collection